server/src/main/java/org/apache/cassandra/sidecar/coordination/CassandraClientTokenRingProvider.java

```diff
@@ -0,0 +1,315 @@
+    {
+        Map<String, Map<Host, Set<TokenRange>>> perKeyspaceTokenRanges = new HashMap<>();
+        for (KeyspaceMetadata ks : metadata.getKeyspaces())
+        {
+            Map<Host, Set<TokenRange>> perHostTokenRanges = new HashMap<>();
+            for (Host host : metadata.getAllHosts())
+            {
+                Set<TokenRange> tokenRanges = metadata.getTokenRanges(ks.getName(), host)
+                                                      .stream()
+                                                      .flatMap(range -> TokenRange.from(range).stream())
+                                                      .collect(Collectors.toSet());
+                perHostTokenRanges.put(host, tokenRanges);
+            }
+            perKeyspaceTokenRanges.put(ks.getName(), perHostTokenRanges);
+        }
+        return perKeyspaceTokenRanges;
+    }
+
+    public static Multimap<CassandraInstance, Range<BigInteger>> calculateTokenRanges(List<CassandraInstance> instances,
+                                                                                      int replicationFactor,
+                                                                                      Partitioner partitioner)
+    {
+        Preconditions.checkArgument(replicationFactor != 0, "Calculation token ranges wouldn't work with RF 0");
+        Preconditions.checkArgument(instances.isEmpty() || replicationFactor <= instances.size(),
+                                    "Calculation token ranges wouldn't work when RF ("
+                                    + replicationFactor + ") is greater than number of Cassandra instances "
+                                    + instances.size());
+        Multimap<CassandraInstance, Range<BigInteger>> tokenRanges = ArrayListMultimap.create();
+
+        for (int index = 0; index < instances.size(); ++index)
+        {
+            CassandraInstance instance = instances.get(index);
+            int disjointReplica = (instances.size() + index - replicationFactor) % instances.size();
+            BigInteger rangeStart = new BigInteger((instances.get(disjointReplica)).token);
+            BigInteger rangeEnd = new BigInteger(instance.token);
+            if (rangeStart.compareTo(rangeEnd) >= 0)
+            {
+                tokenRanges.put(instance, Range.openClosed(rangeStart, partitioner.maximumToken().toBigInteger()));
+                if (!rangeEnd.equals(partitioner.minimumToken().toBigInteger()))
+                    tokenRanges.put(instance, Range.openClosed(partitioner.minimumToken().toBigInteger(), rangeEnd));
+            }
+            else
+            {
+                tokenRanges.put(instance, Range.openClosed(rangeStart, rangeEnd));
+            }
+        }
+        return tokenRanges;
+    }
+
+    /**
+     * Class to encapsule Cassandra instance data
+     */
+    private static class CassandraInstance
```

Comment:
It is used by `protected` methods. To allow those method to remain use-able, you need to open the access to this class to at least `protected`